Add Stream module

tyru opened the pull request:
Add 'Stream' module, that the API is made to resemble Java 8 Stream API.

# Purpose

This module provides:

* **Laziness** like Data.LazyList
* **chaining** like [underscore.vim](https://github.com/haya14busa/underscore.vim)
* Functional interface focusing what you want than how you solve a problem
  * Stream module builds execution plan instead of you
* An interface of stream generation function like Java's Spliterator.
  See generator() document
* A higher-order functions that support string expression, funcref, Data.Closure
  * String expression: `s:Stream.of(1,2,3).map('v:val * 2')`
  * Funcref (lambda): `s:Stream.of(1,2,3).map({n -> n * 2})`
  * Data.Closure: `s:Stream.of(1,2,3).map(s:Closure.from_expr('a:1 * 2'))`

# Example

https://gist.github.com/tyru/d48ad6f7e5b0db45f187606b10b6f42b

```viml
" Random generator (linear congruential generators)
function! s:make_rand() abort
  let seed = reltime()[1]
  let max = float2nr(pow(2, 31) - 1)
  " rand generates random numbers.
  " skip(): the first number seems too small...
  " map(): limits to 0.0 <= val <= 1.0
  " distinct(): gets rid of the same second decimal place of numbers
  return s:Stream.iterate(seed, '(48271 * v:val) % '.max)
                \.skip(1)
                \.map('v:val / '.max.'.0')
                \.distinct('float2nr(round(v:val * 100))')
endfunction

" echo 20 random numbers
for d in s:make_rand().limit(20).to_list()
  echo d
endfor
```

# Limitation

A stream is not reusable as same as Java Stream API.
This limitation enables internal optimization in the point of view of immutability.

```viml
" This throws 'vital: Stream: stream has already been operated upon or closed'
let s = Stream.of([1,2,3,4,5])
call s.map('v:val + 1').to_list()
call s.map('v:val + 1').to_list()
```

# TODO

- Support interface of iterator function like Java's `Spliterator`
  - ~Add `Stream.iterator(dict)` and `dict` is Dictionary which has:~
    - ~`dict.has_next(times)`~
    - ~`dict.next(times)`~
    - `Stream.generator(dict)` is simpler version
  - [x] Add `Stream.generator(dict)` and `dict` is Dictionary which has:
    - `dict.yield(times, NONE)`: returns `NONE` if no more values. otherwise returns a value of stream
- higher-order functions like Stream.map(), Stream.filter() takes String (expression containing v:val), Funcref, Data.Closure
  - [x] iterate(init, func)
  - [x] generate(func)
  - [x] Stream.peek(func)
  - [x] Stream.map(func)
  - [x] Stream.flatmap(func)
  - [x] Stream.filter(func)
  - [x] Stream.slice_before(func)
  - [x] Stream.foreach(func)
  - [x] Stream.take_while(func)
  - [x] Stream.drop_while(func)
  - [x] Stream.sorted([comparator])
  - [x] Stream.reduce(func [, init])
  - [x] Stream.max_by(func [, default])
  - [x] Stream.min_by(func [, default])
  - [x] Stream.find(func [, default])
  - [x] Stream.any(func)
  - [x] Stream.all(func)
  - [x] Stream.none(func)
  - [x] Stream.to_dict(key_mapper, value_mapper [, mergefunc])
  - [x] Stream.group_by(func)
- ~Use Data.Optional in some methods~ Throw exception in some methods if stream is empty and default argument was not given
  - [x] Stream.max([default])
  - [x] Stream.max_by(func [, default])
  - [x] Stream.min([default])
  - [x] Stream.min_by(func [, default])
  - [x] Stream.first([default])
  - [x] Stream.last([default])
  - [x] Stream.find(func [, default])
  - [x] Stream.get_comparator([default])
- ~Throw exception if upstream is infinite stream~ It cannot determine if infinite stream stops! 22be22deac4dba4903fb593cc8c0b7c4b56fdcea
  - sorted([comparator])
  - to_list()
  - collect()
  - reduce(func [, init])
  - average()
  - sum()
  - max()
  - max_by()
  - min()
  - min_by()
  - ~count([func])~
    - It returns 1/0
  - ~foreach()~
    - In general, users may want to do a process with side-effects when calling foreach(), or may want infinite loop.

## Functions

- [x] of(list)
- [x] chars(str [, characteristics])
- [x] lines(str [, characteristics])
- [x] from_list(list [, characteristics])
- [x] from_dict(dict [, characteristics])
- [x] empty()
- [x] zip(stream1, stream2, ...)
- [x] concat(stream1, stream2, ...)
- [x] iterate(init, func)
- [x] generate(func)
- ~range(start_inclusive, end_exclusive)~
- ~range_closed(start_inclusive, end_inclusive)~
- [x] range({expr} [, {max} [, {stride}]])
  - same as Vim script's `range()`

## Methods

- [x] Stream.has_characteristics(characteristics)
- [x] Stream.get_comparator([default])

- [x] Stream.zip(stream, ...)
- [x] Stream.zip_with_index()
- [x] Stream.concat(stream, ...)
- [x] Stream.peek(func)
- [x] Stream.map(func)
- [x] Stream.flatmap(func)
- [x] Stream.filter(func)
- [x] Stream.slice_before(func)
- [x] Stream.foreach(func)
- [x] Stream.skip(n)
- [x] Stream.limit(n)
- [x] Stream.take_while(func)
- [x] Stream.drop_while(func)
- [x] Stream.distinct([stringifier])
- [x] Stream.sorted([comparator])

- [x] Stream.to_list()
- [x] Stream.count([func])
- ~Stream.collect(collector)~
  - maybe unnecessary because there are `to_list()`, `foreach()`, or any other terminal operations
- ~Stream.collect(supplier, accumulator)~
  - maybe unnecessary because Vim script does not have so many data types like Java containers
- [x] Stream.reduce(func [, init])
- [x] Stream.average()
- [x] Stream.sum()
- [x] Stream.max([default])
- [x] Stream.max_by(func [, default])
- [x] Stream.min([default])
- [x] Stream.min_by(func [, default])
- [x] Stream.first([default])
- Stream.find_any([default])
  - same as first() in this module because Vim script does not have thread
- [x] Stream.find(func [, default])
  - same as `filter(func).limit(1).first(default)`
- [x] Stream.last([default])
- [x] Stream.any(func)
- [x] Stream.all(func)
- [x] Stream.none(func)

- ~Collectors.to_list()~
  - unnecessary. because Stream has already Stream.to_list().
- ~Collectors.to_dict(key_mapper, value_mapper [, mergefunc])~
- [x] Stream.to_dict(key_mapper, value_mapper [, mergefunc])
  - if stream is 'DISTINCT', duplication check is not performed
- ~Collectors.joining([sep])~
- [x] Stream.string_join([sep])
  - ~if no sep was given, joined by empty string~
  - if no sep was given, joined by " " (one whitespace) as well as Vim script's `join()`)
- ~Collectors.partitioning_by(func)~
- Stream.partition_by(func)
  - maybe unnecessary because Stream.group_by(func) is generalized version?
- ~Collectors.grouping_by(func)~
- [x] Stream.group_by(func)

# TODO (not implemented in this PR)

## Functions

- from_channel(channel [, characteristics])

## Methods

- Comparator.comparing(prop, is_func)
- Comparator.comparing(prop, is_func).thenComparing(prop, is_func)
- Comparator.comparing(prop, is_func).reverse()

thinca commented:
How about using `Data.Closure`?

lambdalisue commented:
Does the following work?

```vim
let foo = 2
call s.map('v:val + foo').to_list()
```

I previously tried to make something similar but I ended up because I could not make the code above available.

tyru commented:
@thinca oh, I forgot Data.Closure support.

@lambdalisue No. to make it available:

* Using Vim 8 and use lambda(closure) or partial
* Support Data.Closure

rhysd commented:
Nice module! I want to see that this module is added to vital.vim.

If `Stream::of` accepts only an array, how about using variadic arguments for `Stream::of`? 

```vim
" Before
let s = Stream.of([1,2,3,4,5])
call s.map('v:val + 1').to_list()

" After
let s = Stream.of(1,2,3,4,5)
call s.map('v:val + 1').to_list()
```

lambdalisue commented:
While Vim has a limitation of the number of argqments, both syntax need to be supported at least. In case :+1:

lambdalisue commented:
I see. Thanks! @tyru

rhysd commented:
I guess, `Stream::from(list)` or `Stream::from_list(list)` should make a stream from list.

I didn't know that max number of arguments of function in Vim script is 20. I feel 20 is sufficient for `Stream::of` because arguments are written by hand.

lambdalisue commented:
20 is not enough sometime while function can be called by call(fn, args). But if there is Stream.from, I'm ok with that :+1:

tyru commented:
> If Stream::of accepts only an array, how about using variadic arguments for Stream::of?

oh, I saw that in my first design API :)

@rhysd @lambdalisue okay, I will change `Stream.of(...)` arguments and add `Stream.from_list(list)`!

rhysd commented:
> oh, I saw that in my first design API :)

You mean that you designed to use variadic arguments for `Stream.of` but later you decided to use an array for `Stream.of` instead? If so, can I ask the reason?

tyru commented:
well, no clear reason...

1. "Java's Stream.of(T... values) is variadic arguments. let's design it as variadic template"
1. "hm, it seems useful to create stream from variable. and in Java, passing each argument / passing array to variadic argument is same effect. it wouldn't be wrong maybe..."
1. now

tyru commented:
Providing both Stream.of(...) and Stream.from_list() should be the right way :smile:

kamichidu commented:
nice module!

> Stream.of()
of() means it's a static factory method in java.
it can be a really simple name, but it depends on infering argument type and overloading.

In my opinion, its module should provides from_list() for a list stream, from_dict() for a items(dict) stream and of() that is a syntax sugar for limited number of variadic arguments since it's just the vim script.

rhysd commented:
And `from_channel` would be nice! 😸

tyru commented:
> Stream.from_channel

Sounds good!
it is like BlockingQueue in Java.
maybe the granularity of input (string) is the problem, though.

tyru commented:
@kamichidu

Thanks for suggestion.
I also think adding construction function from various data type like from_dict().
because this module has 'characteristics' (same as Java) like stream is ORDERED, SORTED, and DISTINCT (keys of dict must be distinct).
More optimization can be achieved to have more many and correct information of stream.

rhysd commented:
This is not so important point... Characteristics flags may be able to be bit flags such as `0x1`, `0x2`, `0x4`, `0x8`, ... It'd be faster to lookup and enable/disable flags.

tyru commented:
Vim 7 has bit operation functions like and() ?

tyru commented:
oh, bit functions were added in 7.4.
http://www.soum.co.jp/misc/v/vim8-no-susume/1.html
vital.vim supports Vim 7.4 or higher.
so I can use bit functions in vital! thanks @rhysd.

tyru commented:
hmm... Java's `IntStream.range(1,10)` generates [1...9].
`IntStream.rangeClosed(1,10)` generates [1...10].
Vim script `range(1,10)` generates [1...10].
maybe confusing, but this module aims compatible API with Java Stream API, `Stream.range(1,10)` generates [1...9].

lambdalisue commented:
> maybe confusing, but this module aims compatible API with Java Stream API

Why do you need to make it compatible? I think it should follows **Vim's flavour** rather than Java's while users use this module in Vim script.

rhysd commented:
I have the same thought. Referencing other language's APIs is good idea because they're already much considered. But making APIs compatible with other language imports the language's way. For example, Java's `Stream` API has `IntStream`, it looks because Java`s primitive types are special. But in Vim script does not have such a restriction. So we can design our APIs freely.

tyru commented:
@lambdalisue @rhysd 
Cc: @thinca 

> Why do you need to make it compatible?

sorry, "compatible" is wrong.
At first, I do not think I need to design this module is compatible with Java.
I want Java Stream API's chaining and laziness.
Others are welcome for suggestion.

> Java's Stream API has IntStream, it looks because Java's primitive types are special. But Vim script does not have such a restriction.

Yes.

> So we can design our APIs freely.

Yes, but how?
I don't think current API is so bad as Vim script API.
what is "Vim script way"?
Please give me a concrete idea.

tyru commented:
Link: http://lingr.com/room/vim/archives/2017/04/29#message-24015381

rhysd commented:
> I don't think current API is so bad as Vim script API.
what is "Vim script way"?

Yeah, I'm not saying current APIs are not good. I just concern that APIs are unnecessarily influenced from Java APIs because I saw the word 'compatible'. By above comment, I understand that you don't mean that. So please consider my above comment seriously.

> Link: http://lingr.com/room/vim/archives/2017/04/29#message-24015381

Discussing in Lingr separately is not good. Let's discuss here.

tyru commented:
Thanks for reply.
yeah, I also think discussing in many places is not good.
and Lingr is chat service not place for discussion.
sorry.

rhysd commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,368 @@
+
+function! s:ORDERED() abort
+  return s:ORDERED
+endfunction
```
Just a question: Why don't you use `s:ORDERED` directly? Is this function necessary?

tyru replied:
because I suppose they are used by users.
users can use these flags by `stream.has_characteristic(Stream.ORDERED())`.

rhysd replied:
I got it. Thank you. (Method `stream.isOrdered()` may be another candidate for this)

tyru replied:
hmm, isn't it useful if it can check two or more than characteristics?
`stream.has_characteristics([Stream.ORDERED(), Stream.SORTED()])`
however, characteristics is rarely checked by users, though...

rhysd replied:
cool, I didn't know that.

tyru replied:
あ、すみません。
伝えたかったニュアンスとしては「2個以上の characteristics をチェックできたら便利（嬉しい）ですかね？」って意味（提案）でした。

rhysd replied:
なるほど，そういう意味でしたか．そこは，どういう時に characteristics をユーザが知りたいのか僕が良く分かっていないので何とも言えないです…すみません

tyru replied:
`has_characteristics({list of flags or flag})` に変えました。 8969325802daa455782405065b7642161b702907

rhysd commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,368 @@
+endfunction
+
+function! s:Stream.reduce(f, ...) abort
+  let Result = get(a:000, 0, 0)
```
Um, default initial value is `0`. `reduce` might have 2 versions, `fold` and `fold1`. `fold` takes initial value from caller. `fold1` yields first element as initial value.

tyru replied:
hm, we should provide two reduce method, `reduce`, `reduce1`?

tyru replied:
umm, I don't see the clear reason why we should provide two reduce.
Do you think it is not good to provide only `reduce()` which takes initial value as optional value as well as JavaScript and Java? (Java is overload, though)

rhysd replied:
> hm, we should provide two reduce method, reduce, reduce1?

なんか英語で書いたせいでニュアンスが伝わっていないみたいなので日本語で書きますが，should とは書いていないです．might です．「もしかしたら〜かもね」ぐらいの意味です．

> umm, I don't see the clear reason why we should provide two reduce.

説明不足ですみません．ここでの reduce って `reduce(pred)` ならストリームの最初の値が初期値として使われ，`reduce(pred, init)` なら `init` が初期値として使われ，`reduce(pred)` かつストリームが空なら `0` が初期値として使われる，ということであってますよね？
議論したほうが良いかなと感じたポイントは下記の2点です

1. この引数の有無で挙動が変わるので初期値をユーザが提供するバージョンとしないバージョンの2通りを別メソッドで提供したほうが良いかな？と感じた点
2. ストリームが空で初期値も与えられていなかった時に初期値として 0 が使われるのが微妙じゃないかなと感じた点（例外を投げるほうが良さそう？）

rhysd replied:
ただ，1. については確かに JavaScript とかでもそうなっていますし，現状のままで行くのも全然アリだと思っています．

tyru replied:
なるほど！すみません、読み間違えました。 >might

reduce() の件、確かに Java でも初期値与えないバージョンは Optional が返るんですよね。
挙動が変わるし確かに分けちゃった方がいいと思いました。
名前は default を持つ方が reduce()、持たない方は reduce_present() とかどうでしょう？

tyru replied:
@rhysd 変えちゃいました。 721fac889787d36008e1c83a3c9d8e59c83a4ad4

rhysd replied:
`reduce_present` はどうだろう…と思いましたが，確かに良い名前を思いつかないですね…（foldl は foldl1 ですし）言っといてなんですが，分けないのも手なのかな…

tyru replied:
自分も実装しといてアレですが、省略すると例外を投げるのは max() とかと同じ挙動なので統一性があって reduce だけ1つメソッドを作るのもなんだか特別扱いしてるようにも思えます。
どうしましょう？

tyru replied:
@rhysd この件自分としてはやはり max() と合わせて reduce(f [, init]) オンリーに変えてしまいたいです。

rhysd commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,368 @@
+
+function! s:Stream.max_by(f, ...) abort
+  if self.__estimate_size__() == 0
+    return get(a:000, 0, 0)
```
If the stream is empty and there is no default value, I feel we should raise an exception instead of `0`.

tyru replied:
Okay.
In Java Stream API, max() returns Optional.
But Vim script has no Optional (depending Data.Optional is little bit worrying about many dependencies... it causes that importing Stream module makes repository fat).
so I also think exception (when no default) is better.

rhysd commented:
I feel `take(n)`, `drop(n)` should be there :+1: (although I don't know how they are represented in Java Stream API)

tyru commented:
In Java Stream API,
take(n) is limit(n).
drop(n) is skip(n)... ;)
I'm going to implement limit(n) and skip(n).
Should I change the method name?

rhysd commented:
> Should I change the method name?

I feel not necessary. `limit` and `skip` look good name :+1:

rhysd commented:
AFAIK, funcrefs' names should be started with upper case. Should we consider that for elements of streams?

tyru commented:
> Should we consider that for elements of streams?

yeah, we should, I think.
Is there some code without capitalizing?

rhysd commented:
I'm not sure but, don't `a:` variables need to start with upper case for funcrefs? (e.g. here: https://github.com/vim-jp/vital.vim/pull/498/files#diff-d3368c76429ab6adb7d013fbf9d9aea9R153 )

tyru commented:
@rhysd ah, arguments are not necessary to be capitalized.
see this: https://gist.github.com/tyru/6f8b74ad9f69dd99df8df028b1c429d7
But `let f = a:f` will raise E704 error.

tyru commented:
and this is safe: `let o = {} | let o.f = a:f`
`o.f` is not capitalized but no error.
Dictionary's key name does not need to be capitalized.

rhysd commented:
Oh, I see. Thank you for detailed comment.

tyru commented:
Added `flatmap(func)`, but this is little bit different with `Stream#flatMap()` of Java Stream API.
The difference is that func returns List, not Stream.
because it is easy to create List in Vim script than Stream **in String function expression.**
like this:

```viml
echo Stream.of(1,2,3).flatmap('repeat([v:val], v:val)') == [1,2,2,3,3,3]
```

but if func returns Stream, need to capture stream variable.

```viml
" g:Stream is global variable. to make it local variable, expression needs to capture local variable
let g:Stream = vital#vital#new().import('Stream')
echo Stream.of(1,2,3).flatmap('Stream.from_list(repeat([v:val], v:val))') == [1,2,2,3,3,3]
```

This requires lambda or Data.Closure.
But taking func which returns Stream may be overkill because
in most cases func just returns finite sequence, not infinite stream.

tyru commented:
Strange...
only AppVeyor seems to fail because .flatmap()'s value was different with expected value.

* (NG) AppVeyor: 8.0.0003, 8.0.0134
* (OK) Travis CI: 8.0.0589
* (OK) My local Vim: 8.0.0586

tyru commented:
I can't reproduce it in my local machine.
Because my Vim is MSYS Vim?

1. checked out v8.0.0134 branch
1. build
1. `export PATH={vimrepo}/src:$PATH`
1. themis

This is all green.

tyru commented:
I can reproduce it same kaoriya Vim as AppVeyor!

1. Download from http://vim-jp.org/redirects/koron/vim-kaoriya/latest/win64/ (current version is 8.0.0134)
1. `set PATH={Downloaded Vim};{Git for Windows};{themis.bat dir};%PATH%`
1. Run `themis.bat >output.txt` (output to cmd.exe is unreadable... why?)
1. .flatmap() test was failed

```
not ok 742 - Stream .flatmap() flattens mapped stream
# Error occurred line:
#   5:       Assert Equals(Stream.of(1,2,3).flatmap('repeat([v:val], v:val)').to_list(),[1,2,2,3,3,3])
# The equivalent values were expected, but it was not the case.
# 
#     expected: [1, 2, 2, 3, 3, 3]
#          got: [1, 2, 2]
```

tyru commented:
I could confirm the cause... https://github.com/vim-jp/issues/issues/1049

tyru commented:
* Changed `Stream.from_chars()` to `Stream.chars()`, because adjustment for `Stream.lines()` method.
`Stream.from_chars()` is different with other `Stream.from_*()` methods in that existence of preprocess.
`Stream.chars()` splits each character and makes stream. but for example `Stream.from_list()` takes plain List and makes stream.
* `Stream.chars()` is different with Java `String#chars()` in that `String#chars()` returns `int of stream` not `character of stream`. But `Stream.chars()` returns `character of stream`.

rhysd commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,756 @@
+  return s:iterate(map([a:f], a:f)[0], a:f)
+endfunction
+
+function! s:zip(s1, s2) abort
```
`zip` takes fixed 2 arguments. How about making number of parameters variable? (implementation may get complicated...)

tyru replied:
interesting! 😄 
I'll try to implement that!

tyru replied:
@rhysd Supported. but feels the code becomes black magic... https://github.com/vim-jp/vital.vim/pull/498/commits/2d42883dd22f4a9921233a96131635314aa71d09#diff-d3368c76429ab6adb7d013fbf9d9aea9R177
(I didn't want to use `:for` because ~of pride~ it slows down processing List)

tyru commented:
I changed sorted() argument. 767317657d7848a8fdaa01a3c4a9dfe0d0754ecc

#### Before

```viml
function! ByDesc(a, b) abort
  return a:a > a:b ? -1 : a:a ==# a:b ? 0 : 1
endfunction
" same arguments as Vim script's sort()
echo Stream.of(2,1,3).sorted('ByDesc').to_list() == [3,2,1]
```

#### After

```viml
" Can pass string expression or funcref
let by_desc = 'v:val[0] > v:val[1] ? -1 : v:val[0] ==# v:val[1] ? 0 : 1'
echo Stream.of(2,1,3).sorted(by_desc).to_list() == [3,2,1]
```

tyru commented:
Changed the top PR comment.
Maybe I would not change it drastically anymore.

rhysd commented:
> Maybe I would not change it drastically anymore.

If it means ready for code reviews, now is the time to add 'Review me' label 👯

tyru commented:
@rhysd ah, thanks. but there are undone TODOs yet.

tyru commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,831 @@
+
+" a:expr is passed to v:val (but it is not meaningless value because
+" a:expr should not have 'v:val')
+function! s:_call_func0_expr(expr, _args) abort
```
@haya14busa vimlint on reviewdog does not ignore leading `_` ? :cry:

tyru commented:
Hi all, I have a question.

Current `all_match(f)` implementation uses `s:_not(a:f)` which returns a function whose value is inverted.
But [`s:_not()` uses string concatenation for funcref](https://github.com/vim-jp/vital.vim/pull/498/commits/d39f9246ed43bc83ebde36791bb1d2bc29a3888e#diff-d3368c76429ab6adb7d013fbf9d9aea9R739), it loses reference relationship. [Test code](https://gist.github.com/tyru/1de2150fa61c11e2f49bfcd24960c276)

```viml
function! s:_not(f, callee) abort
  let type = type(a:f)
  if type is s:t_func
    return '!' . string(a:f) . '(v:val)'
  elseif type is s:t_string
    return '!map([v:val], ' . string(a:f) . ')[0]'
  else
    " ...
  endif
endfunction

function IsEven(n)
  return a:n % 2 == 0
endfunction

echo s:_not('1 == 1') == '!(1 == 1)
echo s:_not(function('IsEven')) == "!function('IsEven')(v:val)"
" 
echo s:_not(function('IsEven')) == "!function('IsEven')(v:val)"
```

### Question

The problem is that `funcref('Func')` is same as `function('Func')`.
Is there a case that is a problem?

tyru commented:
Add `string_join([sep])`.
it is different with Java's `stream.collect(Collectors.joining())` in that default `sep` value is `" "` (one whitespace) not empty string.
because Vim script's default sep value of `join({list [, {sep}])` is one whitespace.

tyru commented:
It's done! Please review! >ALL

tyru commented:
however, maybe reviewdog still warns about `EVL103: unused argument a:_args`.
but that's vimlint config issue, I want reviewdog to change its config ... 😢

tyru commented:
@haya14busa and @rhysd told me the way to suppress vimlint on Twitter. thanks!

haya14busa commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,949 @@
+" a:expr should not have 'v:val')
+" @vimlint(EVL103, 1, a:args)
+function! s:_call_func0_expr(expr, args) abort
+  return map([a:expr], a:expr)[0]
```
`map([a:args], a:expr)[0]` ?

haya14busa replied:
oh, never mind. I misread it.

tyru replied:
hmm, maybe `eval(a:expr)` is more readable.

tyru commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,949 @@
+  return map([a:args], a:expr)[0]
+endfunction
+
+function! s:_get_present_list_or_throw(stream, size, default, callee) abort
```
function name is not correct.
because this function does "Get non-empty list of possible a:size elements. if empty, throw exception if a:default is s:NONE. but if a:default is not s:NONE, return a:default"
`s:_get_present_list_or_default()` is better.

tyru replied:
no, `s:_get_non_empty_list_or_default()` is better.

tyru commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,949 @@
+  return a:closure.call(a:args[0])
+endfunction
+
+" List of one element is passed to v:val
```
doubt. a:args[0] is passed to v:val. not List.

tyru commented:
I have done 1st review.
and I realized that there is no document... :cry:

tyru commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,949 @@
+      for l:Value in list
+        if !has_key(dup, l:Value)
+          let uniq_list += [l:Value]
+          let dup[l:Value] = 1
```
it may be useful if `s:distinct()` can receive function which unique by. default is string().

rhysd commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,990 @@
+let s:save_cpo = &cpo
```
You should add header

tyru replied:
added

rhysd commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,990 @@
+let s:t_bool = 6
+let s:t_none = 7
+let s:t_job = 8
+let s:t_channel = 9
```
Constant names should be UPPER_CASE as other constants like `s:NONE`.

tyru replied:
changed `%s/\<t_\w\+\>/\U\0\E/g`

rhysd commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,990 @@
+
+" function! s:NONNULL() abort
+"   return s:NONNULL
+" endfunction
```
Vim script basically doesn't have null value. Should this comment out'd function remain?

tyru replied:
I think so too.
but I want to leave [above comment of `let s:NONNULL = 0x10`](https://github.com/vim-jp/vital.vim/pull/498/files#diff-d3368c76429ab6adb7d013fbf9d9aea9R51) for comparison with Java Stream API... 🤔

tyru commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,949 @@
+  return s:_zip_characteristics([result] + others)
+endfunction
+
+function! s:concat(s1, s2) abort
```
it is useful if `s:concat()` can take optional streams as well as `s:zip()` !

rhysd commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,990 @@
+        if !has_key(dup, key)
+          let uniq_list += [l:Value]
+          let dup[key] = 1
+        endif
```
I think we need to `unlet l:Value` here because elements in stream may be differently typed like `s:Stream.from(1, 'a')`.

tyru replied:
oh... I searched all `l:[A-Z]\w*` words and add `unlet` 🤔

rhysd replied:
Um, diff may be updated while I'm reviewing... Please ignore if not needed.

rhysd commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,990 @@
+  let stream._upstream = self
+  let stream.__end = 0
+  " see stream.__take_possible__()
+  " let stream.__sorted_list = []
```
Please describe more. I could not understand what this comment is saying.

tyru replied:
added

rhysd commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,990 @@
+endfunction
+
+
+let s:Stream = {}
```
We may be able to specify some common default properties (e.g. `"__end": 0`)

tyru replied:
`__end` is not in `s:_inf_stream()` 🤔

rhysd commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,990 @@
+  return a:0 ? s.find_first(a:1) : s.find_first()
+endfunction
+
+function! s:Stream.any_match(f) abort
```
How about simply naming `any`? 'match' may be confusing because the word is used for regex commonly.

tyru replied:
cool! I'll change it 😄

tyru replied:
changed

rhysd commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,990 @@
+  return self.filter(a:f).find_first(s:NONE) isnot s:NONE
+endfunction
+
+function! s:Stream.all_match(f) abort
```
How about simply naming `all`? 'match' may be confusing because the word is used for regex commonly.

tyru replied:
changed

rhysd commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,990 @@
+  return self.filter(s:_not(a:f, 'all_match()')).find_first(s:NONE) is s:NONE
+endfunction
+
+function! s:Stream.none_match(f) abort
```
How about simply naming `none`? 'match' may(ry

tyru replied:
ch(ry

rhysd commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,990 @@
+  return self.reduce('v:val[0] + v:val[1]', 0) / n
+endfunction
+
+function! s:Stream.count() abort
```
How about taking an optional predicate? For example, `s:Stream.from(1,2,3,4).count('v:val % 2 == 0')` is `2`.

tyru replied:
cool! I'll add it.

tyru replied:
added

rhysd commented on test/Stream.vimspec:
```diff
@@ -0,0 +1,1641 @@
+      unlet generator
+    End
+  End
+End
```
Please add mode line like `vim: ft=vim`. GitHub will detect it and highlight code as Vim script!

tyru replied:
this is not written in other specs, but I think it is good idea.
added.

rhysd replied:
I guess all specs should add mode line. Now GitHub highlights correctly (and page load is a bit heavier)!

rhysd commented on test/Stream.vimspec:
```diff
@@ -0,0 +1,1641 @@
+    End
+  End
+
+  Describe .of()
```
Let's add tests for heterogeneous stream like `Stream.of(1, 'aaa', 3.14)`

rhysd commented on test/Stream.vimspec:
```diff
@@ -0,0 +1,1641 @@
+  End
+
+  Describe .lines()
+    It constructs stream of each line from String
```
How about adding test cases for `\r\n`?

tyru replied:
oops! added the tests and it failed... fixed. thanks!

rhysd commented on test/Stream.vimspec:
```diff
@@ -0,0 +1,1641 @@
+      Assert Equals(s.has_characteristics(Stream.SORTED()), 1)
+      Assert Equals(s.has_characteristics(Stream.SIZED()), 1)
+      Assert Equals(s.has_characteristics(Stream.IMMUTABLE()), 1)
+      unlet s
```
Is this `unlet s` needed? It looks local variable.

tyru replied:
removed.

rhysd commented on test/Stream.vimspec:
```diff
@@ -0,0 +1,1641 @@
+
+  Describe .from_dict()
+    It constructs stream from Dictionary
+      Assert Equals(Stream.from_dict({}).count(), 0)
```
Let's add the test cases where types of values in dict are different like `{"foo": 42, "bar": 3.14}`

tyru replied:
added

rhysd commented on test/Stream.vimspec:
```diff
@@ -0,0 +1,1641 @@
+  Describe .has_characteristics()
+    It checks if the stream has certain characteristic
+      let s = Stream.from_list([], Stream.ORDERED() + Stream.DISTINCT() + Stream.SORTED() + Stream.SIZED() + Stream.IMMUTABLE())
+      Assert Equals(s.has_characteristics(Stream.ORDERED()), 1)
```
nit: IMO, checking true/false should be done with `Assert True` and `Assert False`. But there are already many `Assert Equals` checks, if you feel it's hard to replace, please ignore this comment.

tyru replied:
replaced. seems it is enough.

* `%s/\<Assert Equals(\(.*\), 1)$/Assert True(\1)/`
* `%s/\<Assert Equals(\(.*\), 0)$/Assert False(\1)/`

tyru replied:
oops, replaced some unrelated text... replaced it macro manually...

rhysd commented on test/Stream.vimspec:
```diff
@@ -0,0 +1,1641 @@
+    End
+  End
+
+  Describe .range()
```
I think we should add test cases where first argument is larger than second one like `Stream.range(10, 3)`

tyru replied:
`Stream.range(1, 0)` exists but certainly larger case (+2 or more) is missing. added

rhysd commented on test/Stream.vimspec:
```diff
@@ -0,0 +1,1641 @@
+    End
+  End
+
+  Describe .peek()
```
What happens if calling `.peek()` for empty stream? I think test case should be added.

tyru replied:
added

tyru commented:
@rhysd fixed all!

tyru commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,1033 @@
+  return s:IMMUTABLE
+endfunction
+
+function! s:of(...) abort
```
of() should raise exception for empty arguments.
empty() should be used to create an empty stream.

rhysd commented:
> I realized that there is no document... 😢

I guess you can borrow many descriptions from Java's API doc.

tyru commented:
> I guess you can borrow many descriptions from Java's API doc.

hm, indeed.
I'll look into that!

rhysd commented:
I want to suggest adding [`slice_before`](https://docs.ruby-lang.org/ja/latest/method/Enumerable/i/slice_before.html) in Ruby. What do you think?

tyru commented:
> I want to suggest adding slice_before in Ruby. What do you think?

interesting!
I'll add it as an intermediate operation like `filter()`,
not a terminal operation like `reduce()`.

tyru commented:
> `slice_before`

@rhysd Is this expected behavior? 3f0abc982a01c18a0657b872e01fb4b9f8619e02

tyru commented:
Link: https://gist.github.com/tyru/d48ad6f7e5b0db45f187606b10b6f42b
Sample code of random number generator! 😄

tyru commented:
I'll change find_first() to first() as well as Ruby.
and add last() method too.

tyru commented:
Added `Example` section to the top comment of PR. and added some features to `Purpose` section.

tyru commented:
Maybe I think I should rename method `limit()` and `skip()` to `take()` and `drop()`
because `take_while()` and `drop_while()` exist.
I'll rename them.

tyru commented:
Fixed all problems and add document.
Please review...! (the 2nd)

rhysd commented:
> @rhysd Is this expected behavior? 3f0abc9

Yes! Perfect! Thank you for implementing it.

rhysd commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,575 @@
+        break
+      endif
+    endwhile
+    let self.__end = !open
```
Is this a bug of vimline? `open` is clearly initialized at line 573.

tyru replied:
yes. I fixed it before. comment is "5 days ago" :)

rhysd replied:
Oh, I see. I overlooked the date

rhysd commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,1172 @@
+    let base = deepcopy(a:base)
+    for trait in a:1
+      call extend(base, deepcopy(trait))
+    endfor
```
You might like to use `map()` rather than `for` statement here: `map('extend(base, deepcopy(v:val))`

rhysd commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,1172 @@
+endfunction
+
+" NOTE: This requires '_upstream'.
+let s:WithBufferred = {'__buffer': []}
```
Bufferred -> Buffered

rhysd commented on doc/vital/Stream.txt:
```diff
@@ -0,0 +1,712 @@
+==============================================================================
+INTRODUCTION				*Vital.Stream-introduction*
+
+*Vital.Stream* is a streaming library, that the API is made to resemble Java 8
```
the API is -> its APIs are

tyru commented on doc/vital/Stream.txt:
```diff
@@ -0,0 +1,712 @@
+Terminal operation			*Vital.Stream-terminal-operations*
+------------------
+
+Terminal operations do not return |Vital.Stream-Stream-object|. And it cannot
```
s/it/the stream/

rhysd commented on doc/vital/Stream.txt:
```diff
@@ -0,0 +1,712 @@
+not aim to be like Java 8 Stream API.
+
+==============================================================================
+EXAMPLES				*Vital.Stream-examples*
```
Adding easier example may be better to understand what this library does. (e.g. integers to FizzBuzz string stream)

rhysd commented on doc/vital/Stream.txt:
```diff
@@ -0,0 +1,712 @@
+
+lines({str} [, {characteristics}])	*Vital.Stream.lines()*
+	Shortcut for `s:Stream.from_list(split(str, '\r\?\n', 1), characteristics)` .
+	but this makes empty stream for empty string.
```
but -> But

or

. but -> , but

rhysd commented on doc/vital/Stream.txt:
```diff
@@ -0,0 +1,712 @@
+
+Some methods are not in Java 8 Stream API and vice versa (skip() -> drop(),
+limit() -> take(), and slice_before() is not in Stream API). This module does
+not aim to be like Java 8 Stream API.
```
This library is inspired by Java's Stream API. So adding link to the API doc would be useful.

rhysd commented on doc/vital/Stream.txt:
```diff
@@ -0,0 +1,712 @@
+<
+iterate({init}, {func})		*Vital.Stream.iterate()*
+	Generates an infinite stream. {init} is the first element of stream.
+	the second element of stream is the value that {init} is applied to
```
the -> The

rhysd commented on doc/vital/Stream.txt:
```diff
@@ -0,0 +1,712 @@
+	the second element of stream is the value that {init} is applied to
+	{func}. And the result is applied to {func} until the end (if
+	downstream limited the number of elements, otherwise it results in
+	an infinite-loop).
```
infinite-loop -> infinite loop

tyru commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,1172 @@
+    let do_break = 0
+    let list = []
+    while !do_break
+      let [r, open] = self._upstream.__take_possible__(self._BULK_SIZE)
```
> Just a question: Why taking elements with _BULK_SIZE size chunk?

`Stream.iterate(1, 'v:val + 1').take_while('v:val < 4').to_list()` must stop.
but it can't know how many elements it can take.

> And the _BULK_SIZE looks to be referenced from nowhere. Should we make a constant for this rather than setting it as property of stream object?

Okay 👍

rhysd commented on doc/vital/Stream.txt:
```diff
@@ -0,0 +1,712 @@
+	echo s:Stream.range(1, very_big_number).take(3).to_list()
+<
+generator({dict})		*Vital.Stream.generator()*
+	Creates an finite/infinite stream from generator object {dict}.
```
an finite/infinite stream -> a stream

rhysd commented on doc/vital/Stream.txt:
```diff
@@ -0,0 +1,712 @@
+	function! dict.yield(times, NONE) abort
+	if a:times >= 3
+	  return a:NONE
+	endif
```
indentation is wrong

rhysd commented on doc/vital/Stream.txt:
```diff
@@ -0,0 +1,712 @@
+	|Vital.Stream-Stream.sorted()| method).  Otherwise, if {default} was
+	given, returns {default}.  Or if {default} was not given, throws an
+	exception.
+	NOTE: If no comparator was passed to sorted() method, the
```
Adding blank line before `NOTE:` sentence would be better to read.

rhysd commented on doc/vital/Stream.txt:
```diff
@@ -0,0 +1,712 @@
+	NOTE: If no comparator was passed to sorted() method, the
+	stream does not have a defined comparator.
+
+Intermediate operation			*Vital.Stream-intermediate-operations*
```
operation -> operations

rhysd commented on doc/vital/Stream.txt:
```diff
@@ -0,0 +1,712 @@
+Intermediate operation			*Vital.Stream-intermediate-operations*
+----------------------
+
+Intermediate operations returns |Vital.Stream-Stream-object|.
```
returns -> return

rhysd commented on doc/vital/Stream.txt:
```diff
@@ -0,0 +1,712 @@
+Stream.peek({func})
+	Peeks elements of stream and does not change the values of elements.
+	This method is useful for debugging. You can pass a function which
+	causes side-effect. {func} is a function of 1 arity.
```
side-effect -> side effect

rhysd commented on doc/vital/Stream.txt:
```diff
@@ -0,0 +1,712 @@
+<
+				*Vital.Stream-Stream.flatmap()*
+Stream.flatmap({func})
+	Creates a flattened stream which each element is applied {func}
```
Creates a flattened stream applying {func} to each element

would be more natural

rhysd commented on doc/vital/Stream.txt:
```diff
@@ -0,0 +1,712 @@
+<
+				*Vital.Stream-Stream.filter()*
+Stream.filter({func})
+	Creates a filtered stream which each element is matched to {func}.
```
Creates a stream filtering elements with {func}

would be more natural.

rhysd commented on doc/vital/Stream.txt:
```diff
@@ -0,0 +1,712 @@
+<
+				*Vital.Stream-Stream.slice_before()*
+Stream.slice_before({func})
+	Slices stream into each element before {func} returns non-zero. {func} is
```
stream -> a stream

rhysd commented on doc/vital/Stream.txt:
```diff
@@ -0,0 +1,712 @@
+Stream.reduce({func} [, {init}])
+	Accumulates each element which is applied to {func}. {func} is a
+	function of 2 arity. If {init} is omitted and stream is empty, an
+	exception is thrown.
```
Please describe the case where {init} is omitted and stream is NOT empty.

tyru replied:
added description and example.

rhysd commented on doc/vital/Stream.txt:
```diff
@@ -0,0 +1,712 @@
+				*Vital.Stream-Stream.max_by()*
+Stream.max_by({func} [, {default}])
+	Returns an element whose value returned by {func} is a maximum value.
+	When the stream is empty, if {default} was given, returns {default}.
```
'When the stream is empty and {default} was given' would be more natural. This 'when ..., if ..., ' pattern is seen in elsewhere. Please check them.

rhysd commented on doc/vital/Stream.txt:
```diff
@@ -0,0 +1,712 @@
+				*Vital.Stream-Stream.string_join()*
+Stream.string_join([{sep}])
+	|join()| each elements with {sep}.  Default value of {sep} is " " (one
+	whitespace) as well as Vim script's `join()`)
```
`join()` should be |join()|

rhysd commented on doc/vital/Stream.txt:
```diff
@@ -0,0 +1,712 @@
+<
+				*Vital.Stream-Stream.group_by()*
+Stream.group_by({func})
+	Shortcut for `stream.to_dict(func, '[v:val]', 'v:val[0] + v:val[1]')` .
```
Adding below text may be better.

> Grouping elements of a stream with {func}.

rhysd commented on test/Stream.vimspec:
```diff
@@ -0,0 +1,2018 @@
+    End
+
+    It constructs heterogeneous stream from arguments
+      Assert Equals(Stream.of('foo', function('tr'), [], {}, 0.1).count(), 5)
```
How about chaining more functions like `.filter` or `.map` or ... because this test can find bugs like missing `unlet` in `for` statement body.

rhysd commented:
Adding at leat one test case with lambda expression (only for Vim8) would be better, although I don't think it causes a problem.

tyru commented:
@rhysd Fixed all!

rhysd commented:
Can anyone review this PR?

haya14busa commented on test/Stream.vimspec:
```diff
@@ -0,0 +1,2023 @@
+      let id = Closure.from_expr('a:1')
+      let len = Closure.from_funcname('len')
+
+      Assert Equals(Stream.empty().to_
```
Hmm... github posted comment to wrong place...?

I mean this line.
https://github.com/vim-jp/vital.vim/pull/498/files/74abc303f9bd7abb121d19250df8f342353dfe46#diff-0d7946cc30d08437fa3704744e1da921R2023

tyru replied:
yeah certainly ft=vimspec and ft=vim is different filetypes. thanks.
to highlight it in github, maybe it is better to prepare .gitattribute and put ft=vim (github doesn't support vimspec syntax...) for *.vimspec...?

haya14busa replied:
.gitattribute seems interesting!

rhysd commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,1170 @@
+"     * in order to take all elements from the stream, pass 1/0
+"   * this function returns '[list, open]'
+"     * 'len(list) <= n'
+"     * callee must not invoke this function after 'open == 0' is returned
```
you're correct. it should be caller, who calls a function. Callee is a function called by caller.

tyru replied:
oops, fixed! thanks!

rhysd commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,1170 @@
+
+function! s:_not(f, callee) abort
+  if s:Closure.is_closure(a:f)
+    return a:f.compose('=!a:1')
```
Closure support is introduced because of [this comment](https://github.com/vim-jp/vital.vim/pull/498#issuecomment-297600261). @thinca, what do you think?

tyru replied:
Data.Closure is necessary to emulate closure function in Vim 7.4.
see [this comment](https://github.com/vim-jp/vital.vim/pull/498#discussion_r115111315).

haya14busa replied:
I mean, Data.Closure can create funcref (.to_function()) and I think Stream can just support funcref, not Data.Closure object.

tyru replied:
If this module doesn't support Data.Closure, to_function() seems a bit redundant?

#### Before

```viml
let s:C = vital#of('vital').import('Data.Closure')
echo s:Stream.of(1,2,3).map(s:C.from_expr('a:1 * 2')).to_list()
```

#### After

```viml
let s:C = vital#of('vital').import('Data.Closure')
echo s:Stream.of(1,2,3).map(s:C.from_expr('a:1 * 2').to_function()).to_list()
```

rhysd replied:
I vote to removing Data.Closure dependency because less dependency is better. From @tyru's example, I  could not find not so much benefit against introducing a new dependency.

rhysd commented on doc/vital/Stream.txt:
```diff
@@ -0,0 +1,737 @@
+
+	echo s:Stream.range(1, 100)
+	            \.map(function('s:fizzbuzz'))
+	            \.to_list()
```
Um, you would be correct. This example was added because I want to see simpler example than below random number generator. Do you have any idea to show the power :muscle: of this module?

haya14busa replied:
It might not be simple, but I wish it has some example which contains expensive calculation in the middle of chain.  Some thing like this:

```vim
echo s:Stream.range(1, 100)
  \.filter(function('s:filter_func'))
  \.map(function('s:expensive_func'))
  \.take(10)
  \.to_list()
" expendive_func will be called at most 10 times.
```

tyru replied:
it's difficult to add a good example...😭

haya14busa commented:
BTW, is "characteristic" feature really needed? I don't understand why Stream module should support s:DISTINCT, s:IMMUTABLE, etc... 

SIZE is important, but I'm not sure why it's represented as "characteristic".

haya14busa commented on doc/vital/Stream.txt:
```diff
@@ -0,0 +1,737 @@
+  See |Vital.Stream.generator()|
+- Higher-order functions that support string expression, |Funcref|,
+  |Vital.Data.Closure|
+  - String expression: `s:Stream.of(1,2,3).map('v:val * 2')`
```
I think it's better to note that we cannot use `v:key`.

tyru commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,1170 @@
+  endif
+  let type = type(a:f)
+  if type is s:T_FUNC
+    return '!' . string(a:f) . '(v:val)'
```
hmm, `closure` function or closure-like object is necessary.
So there are 2 solutions here:

* Create Data.Closure object
* Create Data.Closure-like object
  * `s:_get_callfunc_for_func1()` needs to support this object

I don't want to support more different types object... so Data.Closure is appropriate one.

haya14busa replied:
```vim
  function! s:Stream.all(f) abort
   return self.map(a:f).filter('!!v:val == 0').first(s:NONE) is s:NONE
  endfunction
```
How about this and remove s:_not() ?

tyru replied:
cool! removed `s:_not()`.

tyru commented:
> BTW, is "characteristic" feature really needed? I don't understand why Stream module should support s:DISTINCT, s:IMMUTABLE, etc...

SIZE, SORTED (sorted()), DISTINCT (distinct()) is used.
but other flags are certainly not used.
I think ORDERED and IMMUTABLE should be removed to confuse library users.

haya14busa commented:
> SORTED (sorted()), DISTINCT (distinct()) is used.

Yeah, It's used, but I mean, is it really important to implement it (especially is it needed for initial implementation?

I know .sorted() and .distinct() will be faster with these characteristics, but it's easily removed by .map(), .zip() or something.

Also, if the user want sorted or distinct stream, the user want it regardless it's characteristic, I guess.
In other words, i don't see why Stream module should provide public API for it like has_characteristic().
You can hide it from users and can use it for optimization internally.

haya14busa commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,1157 @@
+
+function! s:Stream.sorted(...) abort
+  if self.has_characteristics(s:SORTED)
+    return self
```
if you pass the (different) comparator, returning self here is not good...?

haya14busa commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,1157 @@
+  endif
+  let stream = s:_new(s:Stream)
+  let stream._characteristics =
+  \ s:DISTINCT + s:SORTED + s:SIZED
```
"SORTED" depends on comparator, add `s:SORTED` is not appropriate here i guess...?

tyru replied:
oh, you found!
yes... this is same as Java.

from http://enterprisegeeks.hatenablog.com/entry/2014/05/07/193604

```java
// not sorted, and not distinct list
List<Integer> list = Arrays.asList(1,2,2,5,4,3);

// Output: 12345
list.stream().sorted().distinct().forEach(System.out::print);
System.out.println();

// Create a stream with SORTED, DISTINCT
Stream<Integer> st = StreamSupport.stream(
        Spliterators.spliterator(list.iterator(), list.size(), Spliterator.SORTED | Spliterator.DISTINCT)
        , false);
// Output: 122543
st.sorted().distinct().forEach(System.out::print);
```

tyru replied:
this is not good behavior to import from Java...
hmm... should I remove characteristics...? hmm...

tyru replied:
> this is not good behavior to import from Java...

I mean, I don't think Java API is not good.
Vital.Stream is different with Java Stream, so the precondition is different.

haya14busa replied:
+1 to remove characteristic. At least we don't need "SORTED"

tyru replied:
* I will remove SORTED and DISTINCT flags.
* I will remove public APIs like has_characteristic(), the 2nd argument of from_list().
* but leave characteristics because it is useful to add information SIZED, or like NUM_ONLY (Number elements only).

tyru replied:
the problem behavior was fixed.

```viml
" Before: [1..100], After: [1..10]
echo s:Stream.range(1, 100).distinct('v:val % 10').to_list()
" Before: [1..100], After: [100..1]
echo s:Stream.range(1, 100).sorted({a,b -> b - a}).to_list()
```

haya14busa replied:
> I will remove SORTED and DISTINCT flags.

It seems you didn't remove these flags completely, can you explain the reason?

tyru replied:
SORTED is for get_comparator(), but for DISTINCT, I've forgotten to remove... 😢 
I will remove this in another PR, okay?

haya14busa commented on doc/vital/Stream.txt:
```diff
@@ -0,0 +1,754 @@
+	{comparator} is a function of 2 arity.
+>
+	" Output: [1,2,3]
+	echo s:Stream.of(1,2,2,3,3).distinct().to_list()
```
example is wrong. it's for .distinct() method.

haya14busa commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,1141 @@
+  return s:_new_from_list([], s:SIZED, 'empty()')
+endfunction
+
+function! s:_new_from_list(list, characteristics, callee) abort
```
:%s/calle/caller/g

haya14busa commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,1141 @@
+  return l:Result
+endfunction
+
+function! s:Stream.max(...) abort
```
Are `.max`, `max_by`, `min` and `min_by` needed in Stream module?
It works only for "Number" stream.

Also, what's the difference with `max(stream.to_list())`?

I think we should remove these methods or provide more efficient implementations.

tyru replied:
max() and min() can pass default value.
and max_by() and min_by() are not Number only.

tyru replied:
hmm, I want to leave these methods.
the reason is unclear though...
These methods can be chained unlike Vim script's max()/min().

rhysd replied:
IMO, `max_by` and `min_by` are needed. As @tyru said, they are not only for numbers because its predicate determines how to compare values.

For current implementation, I agree that `stream.min()` is not needed.
However, originally `stream.max()` is different from `max(stream.to_list()`. `stream.max()` does not need to create a list (just need to 'accumulate' max value). So it should be more efficient than latter.

tyru replied:
okay, I will remove max()/min() (democracy wins! :smirk:).
however, I also think max_by()/min_by() are necessary as @rhysd said.
so leave max_by()/min_by() methods.

haya14busa replied:
Yes. If you want to provide max and min, we should implement it by reduce instead of creating list.
I'm happy to review it but I guess we should separate p-r ;)

As for max_by and min_by, it's not so difficult for users to use reduce to implement it if they want.
(And current implementation seems not efficient.)
 The same is true of .sum, .average and .string_join.
They are not core feature of stream and users can implement them easily.
You said we can use "chaining", but I assume these methods are used at the end and won't be chained.

Also, i don't think theses methods are enough for users to write a plugin in many cases, so I concerned about we may ends up with implementing random utility methods to Stream. 
Why does Stream module provide ".average" and not provide ".median"???

I prefer removing small random utilities methods listed above if it's ok for you. But I don't strongly disagree with it.

haya14busa replied:
underscore.vim provides a way to extend methods, so I think it might be better to provide them similar way.

rhysd replied:
> I'm happy to review it but I guess we should separate p-r ;)

I agree. Let's leave `.max` and `.min` to the next steps :+1: About things like `.average` or `.sum`, I don't have any opinion.

tyru replied:
> As for max_by and min_by, it's not so difficult for users to use reduce to implement it if they want.

At first I tried to implement them by reduce, but Vim is crashed... 068041207067554506103c6b27d38b7d0a6590c8 12cc6885b5f821729c963c22fc7306e22ad56baa
But if we implement it now, it may not be crashed...

> I'm happy to review it but I guess we should separate p-r ;)

👍 

> underscore.vim provides a way to extend methods, so I think it might be better to provide them in a similar way.

Interesting. but this PR has bloated so it is better to discuss about new features in another PR.

haya14busa replied:
> At first I tried to implement them by reduce, but Vim is crashed... 0680412 12cc688
But if we implement it now, it may not be crashed...

btw, I think using `a:f` in string was wrong.

tyru commented:
@haya14busa @rhysd fixed all! (thanks for review many times...!)

tyru commented:
sorry, forgot to push...

haya14busa commented on test/Stream.vimspec:
```diff
@@ -0,0 +1,1730 @@
+      return a:n % 2 == 0
+    endfunction
+    function! Answer() abort
+      return 42
```
Answer to the Ultimate Question of Life, the Universe, and Everything! ✨

haya14busa commented on test/Stream.vimspec:
```diff
@@ -0,0 +1,1730 @@
+      Assert Equals(
+      \ Stream.generate(function('Answer')).take(5).to_list(),
+      \ [42, 42, 42, 42, 42])
+    End
```
[nit] want test funcref which returns different value for each call.

haya14busa commented on test/Stream.vimspec:
```diff
@@ -0,0 +1,1730 @@
+  Describe .take()
+    It raises negative number argument
+      try
+        call Stream.of(1,2,3,4,5).take(-1)
```
ditto (use :Throw)

haya14busa commented on test/Stream.vimspec:
```diff
@@ -0,0 +1,1730 @@
+        Assert 1
+      endtry
+      try
+        call Stream.of(1,2,3,4,5).drop(0)
```
Just `call Stream.of(1,2,3,4,5).drop(0)` should be enough. Add comment if you want.

haya14busa commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,1094 @@
+      \     . self._caller
+    endif
+    " max(): fix overflow
+    let n = max([self.__index + a:n, a:n])
```
Question:

Shouldn't it be `min([self.__index + a:n, INT_MAX])` ?

tyru replied:
> Question:
> Shouldn't it be min([self.__index + a:n, INT_MAX]) ?

yes 😭

tyru replied:
ah, it returns negative value when overflow.
`self.__index + a:n < a:n ? 1/0 : self.__index + a:n` may be correct?

tyru commented:
Some discussions are not over yet,
but this PR has too bloated so I want to make another PR for all members here... 😭 
Okay???

haya14busa commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,1094 @@
+" @assert a:args[2] != 0
+" @assert len(a:args) >= 3
+" @assert a:index >= 0
+function! s:_estimate_range_size(args, index) abort
```
I think "estimate" is not appropriate term here. we can get "size" deterministically. No estimation.

`s:_range_size` may be enough.

haya14busa commented on autoload/vital/__vital__/Stream.vim:
```diff
@@ -0,0 +1,1094 @@
+        break
+      endif
+      if stream.__estimate_size__() > 0
+        let list += stream.__take_possible__(a:n - len(list))[0]
```
if we substruct the number of take elements from a:n, we don't need __read__too_much and make it simpler, i guess. What do you think?

tyru replied:
no, it is necessary.
`__read_too_much` is a buffer for the next call.
`s:concat()` needs to store took elements which it read too much.

haya14busa replied:
Test will pass with below patch.
I mean, we should not "read too much" in the first place.
Can you add tests which has to read too much, if i'm wrong?

```diff
diff --git a/autoload/vital/__vital__/Stream.vim b/autoload/vital/__vital__/Stream.vim
index 59d412d6..1f783f06 100644
--- a/autoload/vital/__vital__/Stream.vim
+++ b/autoload/vital/__vital__/Stream.vim
@@ -287,10 +287,6 @@ function! s:concat(s1, s2, ...) abort
     endif
     " concat buffer and all streams
     let list = []
-    if !empty(self.__read_too_much)
-      let list = s:_slice(self.__read_too_much, 0, a:n - 1)
-      let self.__read_too_much = s:_slice(self.__read_too_much, a:n)
-    endif
     for stream in self._upstream
       if len(list) >= a:n
         break
@@ -299,10 +295,6 @@ function! s:concat(s1, s2, ...) abort
         let list += stream.__take_possible__(a:n - len(list))[0]
       endif
     endfor
-    if len(list) > a:n
-      let self.__read_too_much = s:_slice(list, a:n)
-      let list = s:_slice(list, 0, a:n - 1)
-    endif
     " if all of buffer length, streams' __estimate_size__() are 0,
     " it is end of streams
     let sizes = [len(self.__read_too_much)] +
```

rhysd commented:
Appveyor failed because of weird reason. But I don't have a permission to restart job...

https://ci.appveyor.com/project/vim-jp/vital-vim/build/1079/job/wc90ufiy62my169p

haya14busa commented:
I couldn't look into the detail for all methods, doc, tests but LGTM except some comments I left. :+1:

Thank you for the hard work. It's really interesting module.

tyru commented on doc/vital/Stream.txt:
```diff
@@ -0,0 +1,684 @@
+	Shortcut for `s:Stream.concat(self, stream1, stream2, ...)` .
+
+				*Vital.Stream-Stream.peek()*
+Stream.peek({func})
```
peeking infinite stream looks interesting and useful!
I'll add it into TODO list.

tyru commented:
@haya14busa fixed except [this comment](https://github.com/vim-jp/vital.vim/pull/498#discussion_r115139622)!

tyru commented:
> 1. Should 'characteristics' be?

It was left only for SIZED flag.
(but maybe I will remove this for another refactoring)

> 1. Should some terminal methods .min, .max, ... be added?

I already removed these methods: 76a3fbd076b8e2d1dbb714a26cea89928c24d2e2

* get_comparator()
* max() (previous commit)
* max_by()
* min() (previous commit)
* min_by()
* string_join()
* sum()
* average()

> 1. Should Data.Closure be supported as dependency?

I removed this in conversation of https://github.com/vim-jp/vital.vim/pull/498#discussion_r115049114 ,
but this should be discussed with @thinca

tyru commented:
> > 1. Should Data.Closure be supported as dependency?
>
> I removed this in conversation of https://github.com/vim-jp/vital.vim/pull/498#discussion_r115049114 ,
> but this should be discussed with @thinca 

Created #503

rhysd commented:
I didn't intend to discuss the points here no more, just a confirmation for next steps.

tyru commented:
@rhysd okay, wait @haya14busa's review, and if it's OK, let's merge.

tyru commented:
@thinca Thanks for review.
I think you may know, as I said at Lingr, this PR is too complicated with lots of comments and commits.
so we will not discuss here anymore (except @haya14busa's review and obvious mistakes like your review at map()'s example).
so TODOs in your comment will be carried over into the next PR / issue.

> If you intend to merge with incompatible changes planned, I suggest you move the module to Experimental.Stream.

Thanks to @rhysd and @haya14busa, no incompatible changes left at now.
So module name will be 'Stream'.

tyru commented:
> Thanks to @rhysd and @haya14busa, no incompatible changes left at now.
> So module name will be 'Stream'.

except #503 (this will be an incompatible change depending the conclusion of the discussion)

haya14busa commented:
I left LGTM once, but I don't think it's good idea to merge this p-r just because it becomes too big even though it has some potential, especially not backward-incompatible problems. (I don't mean Data.Closure here. I mean it might be better to change .zip() and .concat() parameters for example.

i'm totally ok to close this p-r and reopen it to clear long long comments and make reviewing easier, but I think you shouldn't ignore @thinca's comments and merge it.

tyru commented:
> i'm totally ok to close this p-r and reopen it to clear long long comments and make reviewing easier, but I think you shouldn't ignore @thinca's comments and merge it.

It seems better.
Closed this PR and created #504.
